man/RNAheat.1
2.1.7

.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.38.2.
.TH RNAHEAT "1" "April 2014" "RNAheat 2.1.7" "User Commands"
.SH NAME
RNAheat \- manual page for RNAheat 2.1.7
.SH SYNOPSIS
.B RNAheat
[\fIoptions\fR]
.SH DESCRIPTION
RNAheat 2.1.7
.PP
calculate specific heat of RNAs
.PP
Reads RNA sequences from stdin and calculates their specific heat in the
temperature range t1 to t2, from the partition function by numeric
differentiation. The result is written to stdout as a list of pairs of
temperature in C and specific heat in Kcal/(Mol*K).
The program will continue to read new sequences until a line consisting of the
single character "@" or an end of file condition is encountered.
.TP
\fB\-h\fR, \fB\-\-help\fR
Print help and exit
.TP
\fB\-\-detailed\-help\fR
Print help, including all details and hidden
options, and exit
.TP
\fB\-\-full\-help\fR
Print help, including hidden options, and exit
.TP
\fB\-V\fR, \fB\-\-version\fR
Print version and exit
.SS "General Options:"
.IP
Below are command line options which alter the general behavior of this
program
.TP
\fB\-\-Tmin\fR=\fIt1\fR
Lowest temperature
.IP
(default=`0')
.TP
\fB\-\-Tmax\fR=\fIt2\fR
Highest temperature
.IP
(default=`100')
.TP
\fB\-\-stepsize\fR=\fIFLOAT\fR
Calculate partition function every stepsize
degrees C
.IP
(default=`1.')
.TP
\fB\-m\fR, \fB\-\-ipoints\fR=\fIipoints\fR
The program fits a parabola to 2*ipoints+1 data
points to calculate 2nd derivatives. Increasing
this parameter produces a smoother curve
.IP
(default=`2')
.TP
\fB\-\-noconv\fR
Do not automatically substitude nucleotide "T"
with "U"
.IP
(default=off)
.SS "Model Details:"
.TP
\fB\-4\fR, \fB\-\-noTetra\fR
Do not include special tabulated stabilizing
energies for tri\-, tetra\- and hexaloop hairpins.
Mostly for testing.
.IP
(default=off)
.TP
\fB\-d\fR, \fB\-\-dangles\fR=\fIINT\fR
How to treat "dangling end" energies for bases
adjacent to helices in free ends and multi\-loops
.IP
(default=`2')
.IP
With \fB\-d2\fR dangling energies will be added for the bases adjacent to a helix on
both sides in any case
.HP
\fB\-d0\fR ignores dangling ends altogether (mostly for debugging).
.TP
\fB\-\-noLP\fR
Produce structures without lonely pairs (helices
of length 1).
.IP
(default=off)
.IP
For partition function folding this only disallows pairs that can only occur
isolated. Other pairs may still occasionally occur as helices of length 1.
.TP
\fB\-\-noGU\fR
Do not allow GU pairs
(default=off)
.TP
\fB\-\-noClosingGU\fR
Do not allow GU pairs at the end of helices
.IP
(default=off)
.TP
\fB\-P\fR, \fB\-\-paramFile\fR=\fIparamfile\fR
Read energy parameters from paramfile, instead of
using the default parameter set.
.IP
A sample parameter file should accompany your distribution.
See the RNAlib documentation for details on the file format.
.TP
\fB\-\-nsp\fR=\fISTRING\fR
Allow other pairs in addition to the usual
AU,GC,and GU pairs.
.IP
Its argument is a comma separated list of additionally allowed pairs. If the
first character is a "\-" then AB will imply that AB and BA are allowed
pairs.
e.g. RNAfold \fB\-nsp\fR \fB\-GA\fR  will allow GA and AG pairs. Nonstandard pairs are
given 0 stacking energy.
.TP
\fB\-e\fR, \fB\-\-energyModel\fR=\fIINT\fR
Rarely used option to fold sequences from the
artificial ABCD... alphabet, where A pairs B,
C\-D etc.  Use the energy parameters for GC (\fB\-e\fR
1) or AU (\fB\-e\fR 2) pairs.
.SH AUTHOR

Ivo L Hofacker, Peter F Stadler, Ronny Lorenz
.SH REFERENCES
.I If you use this program in your work you might want to cite:

R. Lorenz, S.H. Bernhart, C. Hoener zu Siederdissen, H. Tafer, C. Flamm, P.F. Stadler and I.L. Hofacker (2011),
"ViennaRNA Package 2.0",
Algorithms for Molecular Biology: 6:26 

I.L. Hofacker, W. Fontana, P.F. Stadler, S. Bonhoeffer, M. Tacker, P. Schuster (1994),
"Fast Folding and Comparison of RNA Secondary Structures",
Monatshefte f. Chemie: 125, pp 167-188


.I The energy parameters are taken from:

D.H. Mathews, M.D. Disney, D. Matthew, J.L. Childs, S.J. Schroeder, J. Susan, M. Zuker, D.H. Turner (2004),
"Incorporating chemical modification constraints into a dynamic programming algorithm for prediction of RNA secondary structure",
Proc. Natl. Acad. Sci. USA: 101, pp 7287-7292

D.H Turner, D.H. Mathews (2009),
"NNDB: The nearest neighbor parameter database for predicting stability of nucleic acid secondary structure",
Nucleic Acids Research: 38, pp 280-282
.SH "REPORTING BUGS"
If in doubt our program is right, nature is at fault.
.br
Comments should be sent to rna@tbi.univie.ac.at.
.SH "SEE ALSO"

RNAfold(1)


NEWS
Some Changes relative to ViennaRNA-1.03

New energy parameters used in the folding routines.
Prediction of minimum free energy structures now includes contributions
from terminal unpaired nucleotides (so called dangling ends).
The partition function treats these contributions approximately.

Energies for GU pairs in are now taken from
L. He et al., Biochemistry 1991, 30, p 11124

Bulges of length 1 now get an additional stacking energy.

The folding routines accept  a list of nonstandard basepairs that should be
allowed to occur. Also, when doing constrained folding, forced pairs will
be honoured even if the pair is usally not allowed (not AU, GC, or GU).
RNAeval should be useful now for structures containing nonstandard pairs.

Some support for constrained folding using partition function: A constraint
of the form i-j must pair or i must not pair will be honoured by disallowing 
all pairs conflicting with the constraint.

Both folding algorthms use less memory and should be a bit faster for long
sequences.

Several smaller bug fixes.


Progs/RNAheat.c
/*
		      Heat Capacity of RNA molecule

		    c Ivo Hofacker and Peter Stadler
			  Vienna RNA package


	    calculates specific heat using C = - T d^2/dT^2 G(T)
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <math.h>
#include "utils.h"
#include "fold_vars.h"
#include "fold.h"
#include "part_func.h"
extern void  read_parameter_file(const char fname[]);

#define PRIVATE      static
#define PUBLIC
#define MAXWIDTH     201

static char rcsid[] = "$Id: RNAheat.c,v 1.9 1998/07/19 17:47:20 ivo Exp $";

PRIVATE float F[MAXWIDTH];
PRIVATE float ddiff(float f[], float h, int m);
PRIVATE void  usage(void);

#define GASCONST 1.98717  /* in [cal/K] */
#define K0 273.15

PRIVATE void heat_capacity(char *string, float T_min, float T_max,
			  float h, int m)
{
   int length, i;
   char *structure=NULL;
   float hc, kT, min_en;
   
   length = strlen(string);
   
   do_backtrack = 0;   

   temperature = T_min -m*h;
   initialize_fold(length);
   structure = (char *) space(length+1);
   min_en = fold(string, structure);
   free(structure); structure=NULL; free_arrays();
   kT = (temperature+K0)*GASCONST/1000;    /* in kcal */
   pf_scale = exp(-(1.07*min_en)/kT/length );
   init_pf_fold(length);
   
   for (i=0; i<2*m+1; i++) {
      F[i] = pf_fold(string, structure);   /* T_min -2h */
      temperature += h;
      kT = (temperature+K0)*GASCONST/1000;
      pf_scale=exp(-(F[i]/length)/kT);
      update_pf_params(length); 
   }
   while (temperature <= (T_max+m*h+h)) {
      
      hc = - ddiff(F,h,m)* (temperature +K0 - m*h -h); 
      printf("%g   %g\n", (temperature-m*h-h), hc);  
      
      for (i=0; i<2*m; i++)
	 F[i] = F[i+1];
      F[2*m] = pf_fold(string, structure); 
      temperature += h;
      kT = (temperature+K0)*GASCONST/1000;
      pf_scale=exp(-(F[i]/length)/kT);
      update_pf_params(length); 
   }
   free_pf_arrays();
}

/* -------------------------------------------------------------------------- */

PRIVATE float ddiff(float f[], float h, int m)
{
   float fp;
   int i;
   float A, B;
   A = (float)(m*(m+1)*(2*m+1)/3 );                            /* 2*sum(x^2) */
   B = (float)(m*(m+1)*(2*m+1) ) * (float)(3*m*m+3*m-1) /15.;  /* 2*sum(x^4) */
   
   fp=0.;
   for (i=0; i<2*m+1; i++)
      fp += f[i]*( A - (float) ( (2*m+1)*(i-m)*(i-m)) );
   
   fp /= ( ( A*A - B*( (float)(2*m+1) ) )*h*h/2. );
   return (float)fp;
   
}

/* -------------------------------------------------------------------------- */

char  scale1[] = "....,....1....,....2....,....3....,....4";
char  scale2[] = "....,....5....,....6....,....7....,....8";

/*----------------------------------------------------------------------------*/

int main(int argc, char *argv[])
{
   char *string, *line;
   char  ns_bases[33]="", *c;
   char  ParamFile[256]="";
   int  i, length, l, sym;
   float T_min, T_max, h;
   int m_points;
   int istty;
   
   T_min=0.; T_max=100.; h=1; m_points=2;
   string=NULL;
   dangles = 2;   /* dangles can be 0 (no dangles) or 2, default is 2 */

   for (i=1; i<argc; i++) {
      if (argv[i][0]=='-') 
	 switch ( argv[i][1] ) {
	  case 'T':
	    if (strncmp(argv[i], "-Tmin", 5)==0) {
	       if (i==argc-1) usage();
	       if (sscanf(argv[++i], "%f", &T_min)==0)
		 usage();
	    }
	    if (strncmp(argv[i], "-Tmax",5)==0) {
	       if (i==argc-1) usage();
	       if (sscanf(argv[++i], "%f", &T_max)==2)
		 usage();
	    }
	    break;
	  case 'h':
	    if (i==argc-1) usage();
	    if (sscanf(argv[++i],"%f",&h)==2)
	      usage();
	    break;
	  case 'n':
	    if ( strcmp(argv[i], "-noGU" )==0) noGU=1;
	    if ( strcmp(argv[i], "-noCloseGU" ) ==0) no_closingGU=1;
	    if ( strcmp(argv[i], "-noLP")==0) noLonelyPairs=1;
	    if ( strcmp(argv[i], "-nsp") ==0) {
	      if (i==argc-1) usage();
	      if (sscanf(argv[++i], "%32s", ns_bases)==0)
		usage();
	    }
	    break;
	  case '4':
	    tetra_loop=0;
	    break;
	  case 'e':
	    if (i==argc-1) usage();
	    if (sscanf(argv[++i],"%d", &energy_set)==0)
	      usage();
	    break;
	  case 'm':
	    if (i==argc-1) usage();
	    if (sscanf(argv[++i],"%d", &m_points)==0)
	      usage();
	    if (m_points<1) m_points=1;
	    if (m_points>100) m_points=100;
	    break;
	  case 'd': dangles=0;
	    break;
	  case 'P':
	    if (i==argc-1) usage();
	    if (sscanf(argv[++i], "%255s", ParamFile)==0)
	      usage();
	    break;
	  default: usage();
	 }
   }

   if (ParamFile[0])
     read_parameter_file(ParamFile);
   
   if (ns_bases[0]) {
     nonstandards = space(33);
     c=ns_bases;
     i=sym=0;
     if (*c=='-') {
       sym=1; c++;
     }
     while (*c) {
       if (*c!=',') {
	 nonstandards[i++]=*c++;
	 nonstandards[i++]=*c;
	 if ((sym)&&(*c!=*(c-1))) {
	   nonstandards[i++]=*c;
	   nonstandards[i++]=*(c-1);
	 }
       }
       c++;
     }
   }
   
   istty = isatty(fileno(stdout))&&isatty(fileno(stdin));
   
   do {
      if (istty) {
	 printf("\nInput string (upper or lower case); @ to quit\n");
	 printf("%s%s\n", scale1, scale2);
      }
      
      if ((line = get_line(stdin))==NULL) break;
      
      /* skip comment lines and get filenames */
      while ((*line=='*')||(*line=='\0')||(*line=='>')) {
	 printf("%s\n", line);
	 free(line);
	 if ((line = get_line(stdin))==NULL) line = "@";
      } 
      
      if (strcmp(line, "@") == 0) break;
      
      string = (char *) space(strlen(line)+1);
      sscanf(line,"%s",string);
      free(line);
      length = strlen(string);
       
      for (l = 0; l < length; l++) string[l] = toupper(string[l]);
      if (istty)
	 printf("length = %d\n", length);
      
      heat_capacity(string, T_min, T_max, h, m_points);
      free(string);
      fflush(stdout);
   } while (1);
   return 0;
}

PRIVATE void usage(void)
{
  nrerror("usage: RNAheat [-Tmin t1] [-Tmax t2] [-h stepsize] [-m ipoints] [-4] [-d]\n"
	  "               [-noGU] [-noCloseGU] [-noLP] [-e 1|2] [-P paramfile] [-nsp pairs]");
}
